Makefile.kokkos

```diff
@@ -321,89 +327,106 @@ KOKKOS_LDFLAGS = -L$(shell pwd)
 KOKKOS_SRC =
 KOKKOS_HEADERS =
 
-# Generating the KokkosCore_config.h file.
 
-tmp := $(shell echo "/* ---------------------------------------------" > KokkosCore_config.tmp)
-tmp := $(shell echo "Makefile constructed configuration:" >> KokkosCore_config.tmp)
-tmp := $(shell date >> KokkosCore_config.tmp)
-tmp := $(shell echo "----------------------------------------------*/" >> KokkosCore_config.tmp)
+gen_config=KokkosCore_config.tmp
+gen_header=KokkosCore_config.h
+# Functions for generating config header file
+wrtkks = echo $1 >> $(gen_config)
 
-tmp := $(shell echo '\#if !defined(KOKKOS_MACROS_HPP) || defined(KOKKOS_CORE_CONFIG_H)' >> KokkosCore_config.tmp)
-tmp := $(shell echo '\#error "Do not include KokkosCore_config.h directly; include Kokkos_Macros.hpp instead."' >> KokkosCore_config.tmp)
-tmp := $(shell echo '\#else' >> KokkosCore_config.tmp)
-tmp := $(shell echo '\#define KOKKOS_CORE_CONFIG_H' >> KokkosCore_config.tmp)
-tmp := $(shell echo '\#endif' >> KokkosCore_config.tmp)
+# Generating the KokkosCore_config.h file.
 
-tmp := $(shell echo "/* Execution Spaces */" >> KokkosCore_config.tmp)
+defines=""
+define1=""
+
+build-config-kokkos-prefix:
+	  @$(shell rm -f $(gen_config))
+	  @$(shell touch $(gen_config))
+	  @$(call wrtkks,"/* ---------------------------------------------")
+	  @$(call wrtkks,"Makefile constructed configuration:")
+	  @$(call wrtkks,$(shell date))
+	  @$(call wrtkks,"----------------------------------------------*/")
+	  
+	  @$(call wrtkks,'#if !defined(KOKKOS_MACROS_HPP) || defined(KOKKOS_CORE_CONFIG_H)')
+	  @$(call wrtkks,'#error "Do not include $(gen_header) directly; include Kokkos_Macros.hpp instead."')
+	  @$(call wrtkks,'#else')
+	  @$(call wrtkks,'#define KOKKOS_CORE_CONFIG_H')
+	  @$(call wrtkks,'#endif')
+	  
+defexes="/* Execution Spaces */"
```

Comment:
again, is it really necessary to have a new system for doing this? and in the same PR?

Reply:
I view this as the most important part of the initial refactor.  Previously these were done with a `echo ... >> Makefile.kokkos` in one section and an `echo ... > kokkos.cmake` in another section.  This then requires a developer to ensure that variables are provided in two separate places.  By using these functions, we only have a single reducing the chances for developer error.  Since the overall goal is to reduce the build system redundancy, this seemed in keeping with that redundancy.

Reply:
> requires a developer to ensure that variables are provided in two separate places. By using these functions, we only have a single reducing the chances for developer error.

okay, that seems like good rationale too

Reply:
I am starting to be ok with this, but check my comments above for naming changes (I will summarize those later in a comment).

Reply:
Can you elaborate on why this need to happen in targets instead of the direct append to files while the Makefile is parsed? It seems to add some complexity for now without apparent reason?

Reply:
Rename: defexes -> KOKKOS_INTERNAL_CONFIG_DEFINES_EXEC